File: sdk/unreal/Source/ChillObservability/Private/ChillRecord.cpp
// Copyright Chill Contributors. Licensed under the Apache License, Version 2.0.

#include "ChillRecord.h"

#include "ChillConfiguration.h"
#include "ChillIds.h"
#include "ChillJson.h"

FString FChillOtlpAttribute::ToJson() const
{
	return ChillJson::Attribute(Key, ValueJson);
}

FString FChillRecord::ToLogJson() const
{
	TArray<FString> EncodedAttributes;
	EncodedAttributes.Reserve(Attributes.Num());
	for (const FChillOtlpAttribute& Attribute : Attributes)
	{
		EncodedAttributes.Add(Attribute.ToJson());
	}

	FString Result = TEXT("{\"timeUnixNano\":");
	Result.Append(ChillJson::Quote(OccurredAtUnixNanoseconds));
	Result.Append(TEXT(",\"observedTimeUnixNano\":"));
	Result.Append(ChillJson::Quote(OccurredAtUnixNanoseconds));
	Result.Append(TEXT(",\"eventName\":"));
	Result.Append(ChillJson::Quote(Name));
	if (!Traceparent.IsEmpty())
	{
		Result.Append(TEXT(",\"traceId\":"));
		Result.Append(ChillJson::Quote(ChillIds::TraceId(Traceparent)));
		Result.Append(TEXT(",\"spanId\":"));
		Result.Append(ChillJson::Quote(ChillIds::SpanId(Traceparent)));
		Result.Append(TEXT(",\"flags\":1"));
	}
	Result.Append(TEXT(",\"attributes\":"));
	Result.Append(ChillJson::Array(EncodedAttributes));
	Result.AppendChar(TEXT('}'));
	return Result;
}

namespace ChillOtlp
{
	const TCHAR* const SchemaUrl = TEXT("https://schemas.chill.dev/behavior/v1/envelope.schema.json");

	TArray<FChillOtlpAttribute> BaseAttributes(
		const FChillConfiguration& Configuration,
		const FChillRecord& Record,
		const FString& SourcePlatform,
		const FString& InstallationId,
		const TArray<FChillAnnotationEntry>& Annotations,
		const TMap<FString, EChillAnnotationClassification>& Definitions)
	{
		TArray<FChillOtlpAttribute> Attributes;
		Attributes.Add(Text(TEXT("chill.schema.version"), TEXT("1.0.0")));
		Attributes.Add(Text(TEXT("chill.schema.url"), SchemaUrl));
		Attributes.Add(Text(TEXT("chill.record.id"), Record.RecordId));
		Attributes.Add(Text(TEXT("chill.subject.id"), Record.SubjectId));
		Attributes.Add(Text(TEXT("chill.behavior.kind"), Record.Kind));
		Attributes.Add(Text(TEXT("chill.behavior.operation"), Record.Operation));
		Attributes.Add(Text(TEXT("chill.behavior.name"), Record.Name));
		Attributes.Add(Integer(TEXT("chill.clock.sequence_number"), Record.Sequence));
		Attributes.Add(Text(TEXT("chill.source.platform"), SourcePlatform));
		Attributes.Add(Text(TEXT("chill.source.installation_id"), InstallationId));
		Attributes.Add(Text(TEXT("chill.context.session_id"), Record.SessionId));
		Attributes.Add(Text(TEXT("chill.privacy.consent"), TEXT("granted")));
		Attributes.Add(Text(TEXT("chill.privacy.policy_version"), Configuration.PolicyVersion));
		Attributes.Add(Text(TEXT("chill.privacy.capture_class"), TEXT("analytics")));
		Attributes.Add(Text(TEXT("chill.privacy.redaction_state"), TEXT("none")));

		for (const FChillAnnotationEntry& Annotation : Annotations)
		{
			const EChillAnnotationClassification* Classification = Definitions.Find(Annotation.Key);
			if (Classification == nullptr)
			{
				continue;
			}
			Attributes.Emplace(
				TEXT("chill.annotation.") + Annotation.Key,
				ChillJson::AnyValue(Annotation.Value));
			Attributes.Add(Text(
				TEXT("chill.privacy.annotation_classification.") + Annotation.Key,
				ChillNames::SnakeCase(*Classification)));
		}
		return Attributes;
	}

	FString Batch(
		const FChillConfiguration& Configuration,
		const FString& SourcePlatform,
		const FString& ProcessId,
		const FString& EngineVersion,
		const FString& OperatingSystem,
		const TArray<FChillQueuedRecord>& Records)
	{
		TArray<FString> Logs;
		Logs.Reserve(Records.Num());
		for (const FChillQueuedRecord& Record : Records)
		{
			Logs.Add(Record.LogJson);
		}

		TArray<FString> Resource;
		Resource.Add(ChillJson::Attribute(TEXT("service.name"), ChillJson::StringValue(Configuration.GetServiceName())));
		Resource.Add(ChillJson::Attribute(TEXT("telemetry.sdk.language"), ChillJson::StringValue(TEXT("cpp"))));
		Resource.Add(ChillJson::Attribute(TEXT("process.runtime.name"), ChillJson::StringValue(TEXT("unreal"))));
		Resource.Add(ChillJson::Attribute(TEXT("process.runtime.version"), ChillJson::StringValue(EngineVersion)));
		Resource.Add(ChillJson::Attribute(TEXT("os.type"), ChillJson::StringValue(OperatingSystem)));
		Resource.Add(ChillJson::Attribute(TEXT("chill.game.engine"), ChillJson::StringValue(TEXT("unreal"))));
		Resource.Add(ChillJson::Attribute(TEXT("chill.source.platform"), ChillJson::StringValue(SourcePlatform)));
		Resource.Add(ChillJson::Attribute(TEXT("chill.source.process_id"), ChillJson::StringValue(ProcessId)));

		return FString(TEXT("{\"resourceLogs\":[{\"resource\":{\"attributes\":"))
			+ ChillJson::Array(Resource)
			+ TEXT("},\"scopeLogs\":[{\"scope\":{\"name\":\"dev.chill.unreal\",\"version\":\"0.1.0\"},")
			+ TEXT("\"schemaUrl\":\"") + SchemaUrl + TEXT("\",\"logRecords\":")
			+ ChillJson::Array(Logs)
			+ TEXT("}]}]}");
	}

	FChillOtlpAttribute Text(const FString& Key, const FString& Value)
	{
		return FChillOtlpAttribute(Key, ChillJson::StringValue(Value));
	}

	FChillOtlpAttribute Boolean(const FString& Key, bool Value)
	{
		return FChillOtlpAttribute(Key, ChillJson::BooleanValue(Value));
	}

	FChillOtlpAttribute Integer(const FString& Key, int64 Value)
	{
		return FChillOtlpAttribute(Key, ChillJson::IntegerValue(Value));
	}

	FChillOtlpAttribute IntegerText(const FString& Key, const FString& Value)
	{
		return FChillOtlpAttribute(Key, ChillJson::IntegerValue(Value));
	}

	FChillOtlpAttribute Number(const FString& Key, double Value)
	{
		return FChillOtlpAttribute(Key, ChillJson::DoubleValue(Value));
	}

	FChillOtlpAttribute Strings(const FString& Key, const TArray<FString>& Values)
	{
		return FChillOtlpAttribute(Key, ChillJson::StringArrayValue(Values));
	}
}

File: sdk/unity/Runtime/ChillRecord.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace Chill.Unity
{
    internal sealed class ChillOtlpAttribute
    {
        internal ChillOtlpAttribute(string key, string valueJson)
        {
            Key = key;
            ValueJson = valueJson;
        }

        internal string Key { get; private set; }
        internal string ValueJson { get; private set; }

        internal string Json
        {
            get { return ChillJson.Attribute(Key, ValueJson); }
        }
    }

    internal sealed class ChillRecord
    {
        internal string RecordId;
        internal string SubjectId;
        internal string Kind;
        internal string Operation;
        internal string Name;
        internal string OccurredAtUnixNanoseconds;
        internal string MonotonicNanoseconds;
        internal long Sequence;
        internal string SessionId;
        internal string Traceparent;
        internal IReadOnlyList<ChillOtlpAttribute> Attributes;

        internal string ToLogJson()
        {
            var all = new List<string>();
            for (int index = 0; index < Attributes.Count; index += 1)
            {
                all.Add(Attributes[index].Json);
            }

            var result = new StringBuilder("{\"timeUnixNano\":");
            result.Append(ChillJson.Quote(OccurredAtUnixNanoseconds));
            result.Append(",\"observedTimeUnixNano\":");
            result.Append(ChillJson.Quote(OccurredAtUnixNanoseconds));
            result.Append(",\"eventName\":");
            result.Append(ChillJson.Quote(Name));
            if (!string.IsNullOrEmpty(Traceparent))
            {
                result.Append(",\"traceId\":");
                result.Append(ChillJson.Quote(ChillIds.TraceId(Traceparent)));
                result.Append(",\"spanId\":");
                result.Append(ChillJson.Quote(ChillIds.SpanId(Traceparent)));
                result.Append(",\"flags\":1");
            }
            result.Append(",\"attributes\":");
            result.Append(ChillJson.Array(all));
            result.Append('}');
            return result.ToString();
        }
    }

    internal sealed class ChillExportBatch
    {
        internal ChillExportBatch(IReadOnlyList<ChillQueuedRecord> records, string body)
        {
            Records = records;
            Body = body;
        }

        internal IReadOnlyList<ChillQueuedRecord> Records { get; private set; }
        internal string Body { get; private set; }
    }

    internal static class ChillOtlp
    {
        private const string Schema = "https://schemas.chill.dev/behavior/v1/envelope.schema.json";

        internal static List<ChillOtlpAttribute> BaseAttributes(
            ChillConfiguration configuration,
            ChillRecord record,
            string sourcePlatform,
            string installationId,
            IReadOnlyList<ChillAnnotationEntry> annotations,
            IReadOnlyDictionary<string, ChillAnnotationClassification> definitions)
        {
            var attributes = new List<ChillOtlpAttribute>
            {
                Text("chill.schema.version", "1.0.0"),
                Text("chill.schema.url", Schema),
                Text("chill.record.id", record.RecordId),
                Text("chill.subject.id", record.SubjectId),
                Text("chill.behavior.kind", record.Kind),
                Text("chill.behavior.operation", record.Operation),
                Text("chill.behavior.name", record.Name),
                Integer("chill.clock.sequence_number", record.Sequence),
                Text("chill.source.platform", sourcePlatform),
                Text("chill.source.installation_id", installationId),
                Text("chill.context.session_id", record.SessionId),
                Text("chill.privacy.consent", "granted"),
                Text("chill.privacy.policy_version", configuration.PolicyVersion),
                Text("chill.privacy.capture_class", "analytics"),
                Text("chill.privacy.redaction_state", "none")
            };

            for (int index = 0; index < annotations.Count; index += 1)
            {
                ChillAnnotationEntry annotation = annotations[index];
                ChillAnnotationClassification classification;
                if (!definitions.TryGetValue(annotation.Key, out classification))
                {
                    continue;
                }
                attributes.Add(new ChillOtlpAttribute(
                    "chill.annotation." + annotation.Key,
                    ChillJson.AnyValue(annotation.Value)));
                attributes.Add(Text(
                    "chill.privacy.annotation_classification." + annotation.Key,
                    ChillNames.SnakeCase(classification)));
            }
            return attributes;
        }

        internal static string Batch(
            ChillConfiguration configuration,
            string sourcePlatform,
            string processId,
            string unityVersion,
            string operatingSystem,
            IReadOnlyList<ChillQueuedRecord> records)
        {
            var logs = new List<string>(records.Count);
            for (int index = 0; index < records.Count; index += 1)
            {
                logs.Add(records[index].LogJson);
            }
            var resource = new List<string>
            {
                ChillJson.Attribute("service.name", ChillJson.StringValue(configuration.ServiceName)),
                ChillJson.Attribute("telemetry.sdk.language", ChillJson.StringValue("csharp")),
                ChillJson.Attribute("process.runtime.name", ChillJson.StringValue("unity")),
                ChillJson.Attribute("process.runtime.version", ChillJson.StringValue(unityVersion)),
                ChillJson.Attribute("os.type", ChillJson.StringValue(operatingSystem)),
                ChillJson.Attribute("chill.game.engine", ChillJson.StringValue("unity")),
                ChillJson.Attribute("chill.source.platform", ChillJson.StringValue(sourcePlatform)),
                ChillJson.Attribute("chill.source.process_id", ChillJson.StringValue(processId))
            };

            return "{\"resourceLogs\":[{\"resource\":{\"attributes\":"
                + ChillJson.Array(resource)
                + "},\"scopeLogs\":[{\"scope\":{\"name\":\"dev.chill.unity\",\"version\":\"0.1.0\"},"
                + "\"schemaUrl\":\"" + Schema + "\",\"logRecords\":"
                + ChillJson.Array(logs)
                + "}]}]}";
        }

        internal static ChillOtlpAttribute Text(string key, string value)
        {
            return new ChillOtlpAttribute(key, ChillJson.StringValue(value));
        }

        internal static ChillOtlpAttribute Boolean(string key, bool value)
        {
            return new ChillOtlpAttribute(key, ChillJson.BooleanValue(value));
        }

        internal static ChillOtlpAttribute Integer(string key, long value)
        {
            return new ChillOtlpAttribute(key, ChillJson.IntegerValue(value));
        }

        internal static ChillOtlpAttribute Integer(string key, string value)
        {
            return new ChillOtlpAttribute(key, ChillJson.IntegerValue(value));
        }

        internal static ChillOtlpAttribute Number(string key, double value)
        {
            return new ChillOtlpAttribute(key, ChillJson.DoubleValue(value));
        }

        internal static ChillOtlpAttribute Strings(string key, IReadOnlyList<string> values)
        {
            return new ChillOtlpAttribute(key, ChillJson.StringArrayValue(values));
        }
    }
}
